planner/core/find_best_task.go

```diff
@@ -700,17 +699,19 @@ func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty,
 	return ts, partialCost
 }
 
-func isCoveringIndex(columns, indexColumns []*expression.Column, idxColLens []int, pkIsHandle bool) bool {
+func (ds *DataSource) isCoveringIndex(columns, indexColumns []*expression.Column, idxColLens []int, tblInfo *model.TableInfo) bool {
+	indexCols := append(indexColumns, ds.commonHandleCols...)
+	indexColLens := append(idxColLens, ds.commonHandleLens...)
 	for _, col := range columns {
-		if pkIsHandle && mysql.HasPriKeyFlag(col.RetType.Flag) {
+		if tblInfo.PKIsHandle && mysql.HasPriKeyFlag(col.RetType.Flag) {
 			continue
 		}
 		if col.ID == model.ExtraHandleID {
 			continue
 		}
 		isIndexColumn := false
-		for i, indexCol := range indexColumns {
-			isFullLen := idxColLens[i] == types.UnspecifiedLength || idxColLens[i] == col.RetType.Flen
+		for i, indexCol := range indexCols {
```

Comment:
We can extract this for loop, and call it twice with `indexCols` and `commonHandleCols`.
So we don't need to append slices.